csm_web/scheduler/models.py

```diff
@@ -30,150 +35,118 @@ class Attendance(models.Model):
+    name = models.SlugField(max_length=100, unique_for_month="enrollment_start")
     valid_until = models.DateField()
     enrollment_start = models.DateTimeField()
     enrollment_end = models.DateTimeField()
 
     def __str__(self):
         return self.name
 
+    def clean(self):
+        super().clean()
+        if self.enrollment_end <= self.enrollment_start:
+            raise ValidationError("enrollment_end must be after enrollment_start")
+        if type(self.enrollment_end) == datetime.datetime:
+            enrollment_end = self.enrollment_end.date()
+        else:
+            enrollment_end = self.enrollment_end
+        if self.valid_until < enrollment_end:
+            raise ValidationError("valid_until must be after enrollment_end")
+
+
+class Profile(ValidatingModel):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+
+    @property
+    def name(self):
+        return f"{self.user.first_name} {self.user.last_name}"
+
+    def __str__(self):
+        if self.section:
+            return f"{self.name} ({self.section.course.name})"
+        return self.name
+
     class Meta:
-        unique_together = ("name",)
-
-
-class Profile(ActivatableModel):
-    STUDENT = "ST"
-    JUNIOR_MENTOR = "JM"
-    ASSOCIATE_MENTOR = "AM"
-    SENIOR_MENTOR = "SM"
-    COORDINATOR = "CO"
-    ROLE_CHOICES = (
-        (STUDENT, "Student"),
-        (JUNIOR_MENTOR, "Junior Mentor"),
-        (ASSOCIATE_MENTOR, "Associate Mentor"),
-        (SENIOR_MENTOR, "Senior Mentor"),
-        (COORDINATOR, "Coordinator"),
-    )
-    ROLE_MAP = dict(ROLE_CHOICES)
+        abstract = True
 
-    leader = models.ForeignKey(
-        "self",
+
+class Student(Profile):
+    section = models.ForeignKey(
+        "Section",
         on_delete=models.CASCADE,
-        related_name="followers",
         blank=True,
         null=True,
-    )
-    course = models.ForeignKey(Course, on_delete=models.CASCADE)
-    role = models.CharField(max_length=2, choices=ROLE_CHOICES, blank=True)
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
-    section = models.ForeignKey(
-        "Section", on_delete=models.CASCADE, blank=True, null=True
+        related_name="students",
     )
 
-    def clean(self):
-        if self.section:
-            mentor_set = self.section._get_mentor_set()
-            if mentor_set.count() > 1:
-                raise ValidationError(_("Cannot save section with more than 1 mentor!"))
-            if self.course != self.section.course:
-                raise ValidationError(_("Profile course must match section course!"))
+    class Meta:
+        unique_together = ("user", "section")
 
-    @property
-    def name(self):
-        return f"{self.user.first_name} {self.user.last_name}"
 
-    def __str__(self):
-        return f"{self.course.name} {self.role} - {self.name}"
+class Mentor(Profile):
+    pass
 
 
-class Section(models.Model):
+class Section(ValidatingModel):
```

Comment:
add a "comment" text field so we can add special notes about a section, like "EOP" or "early start" or whatever

Reply:
I worry about this being abused in place of building new features. EOP I understand (see above comment about `public` field on Section in place of this), but what other purposes would this serve?

Reply:
i think it's better to keep the system as simple as possible, and if we don't need to add additional fields to serve a purpose we shouldn't - my idea is that the "comment" field shouldn't be used to control any logic, just as a way of displaying some extra information

i can see how it could be potentially abused; that said, i think django is easy enough to modify that if we see it getting out of control, we would start feature work pretty quickly - in general, i don't think we should spend more time engineering new solutions than is really necessary

Reply:
Will add